pkg/queue/concurrency_state.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2021 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queue
+
+import (
+	"net/http"
+
+	"go.uber.org/zap"
+)
+
+// ConcurrencyStateHandler does stuff //TODO what stuff?
+func ConcurrencyStateHandler(opts ConcurrencyStateOps) http.HandlerFunc {
```

Comment:
I think if we're passing an opts struct to a Handler method like this I'd probably implement ServeHTTP on the struct instead of passing it in. An example of that is [how timeToFirstByteHandler works](https://github.com/knative/serving/blob/d5d489c0babd5c9ae9a37d707c06df92d70711f5/pkg/http/handler/timeout.go#L31). This would mean we'd use fields rather than closure variables, which may be better anyway.

I'd also be tempted to just drop the ConcurrencyStateOps struct for now, and refactor to a struct with ServeHTTP later. If we _do_ keep it though, it should be `ConcurrencyStateOpts` with a `t` :).

Reply:
Will refactor another day :)